### PR TITLE
feat: modularity

### PR DESCRIPTION
**Pull request by thommann**

### CODE REVIEW

File: Demos/Priority/Allocator/PriorityAllocator.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+from ..Owners.PrioritySpaceOwner import PrioritySpaceOwner
+
+
+class PriorityAllocator(Allocator):
+    @staticmethod
+    def compatible_owner():
+        return [PriorityPathOwner, PrioritySpaceOwner]
+
+    def __init__(self):
+        super().__init__()
```
Comment (johannschwabe) [inline Demos/Priority/Allocator/PriorityAllocator.py]: needed?

File: Simulator/Agents/Agent.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    from ..Segments.Segment import Segment
+    from ..Simulator import Simulator
+    from ..Bids.Bid import Bid
+
+
+class Agent(ABC):
+    agent_type: str
+
+    def __init__(self,
+                 simulator: "Simulator",
```
Comment (johannschwabe) [inline Simulator/Agents/Agent.py]: Agents should not need a simulator instance

File: Simulator/Agents/PathAgent.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if len(self.allocated_segments) > 0 and self.allocated_segments[-1].same(path_segment):
+            self.allocated_segments[-1].join(path_segment)
+        else:
+            self.allocated_segments.append(path_segment)
+
+    def get_allocated_coords(self) -> List["Coordinate4D"]:
+        return [coord for path_segment in self.allocated_segments for coord in path_segment.coordinates]
+
+    def does_collide(self, other_coordinate: "Coordinate4D", other_agent: "PathAgent"):
+        for coordinate in self.get_allocated_coords():
```
Comment (johannschwabe) [inline Simulator/Agents/PathAgent.py]: Iterating over all allocated coordinates is not necessary: using other_coordinate.t the location of interesting coordinates can be directly found

File: Simulator/Agents/PathAgent.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            if other_coordinate.t <= coordinate.t <= other_coordinate.t + other_agent.speed:
+                distance = coordinate.inter_temporal_distance(other_coordinate)
+                if distance == 0:
```
Comment (johannschwabe) [inline Simulator/Agents/PathAgent.py]: für den fall das der near_radius 0 ist?

**Review by johannschwabe — CHANGES_REQUESTED** [review]
some small changes you might have already done

File: Simulator/Agents/PathAgent.py (hunk 4 above)
Comment (thommann) [reply Simulator/Agents/PathAgent.py]: yes

File: Simulator/Agents/PathAgent.py (hunk 3 above)
Comment (thommann) [reply Simulator/Agents/PathAgent.py]: how?

**Review by johannschwabe — APPROVED** [review]

**Pull request merged**